### PR TITLE
Import some fixes for python 3

### PR DESCRIPTION
**Pull request by jhford**
Here's some work towards fixing Python 3 support.  This should work with
python 2 as well.  I tried to avoid using 'six' so we didn't add a dep.

I couldn't get the spread tests to work.  I also do not know how to make
'tox' run both the py35 and py25 environment tests...

### CODE REVIEW

**Comment by jhford** [comment]
Note that this does not fix #2 , it's just a start towards it.

File: slugid/slugid.py [hunk 1]
```diff
@@ -16,7 +18,10 @@ def decode(slug):
     """
     Returns the uuid.UUID object represented by the given v4 or "nice" slug
     """
-    return uuid.UUID(bytes=base64.urlsafe_b64decode(slug + '==')) # b64 padding
+    if not two and isinstance(slug, bytes):
+      slug = slug.decode('ascii')
+    slug = slug + '=='
+    return uuid.UUID(bytes=base64.urlsafe_b64decode(slug)) # b64 padding
```
Comment (petemoore) [inline slugid/slugid.py]: the `# b64 padding` needs to move to line 23

File: slugid/slugid.py [hunk 2]
```diff
@@ -38,6 +43,6 @@ def nice():
     Potentially other "nice" properties may be added in future to further
     restrict the range of potential uuids that may be generated.
     """
-    rawBytes = uuid.uuid4().bytes
-    rawBytes = chr(ord(rawBytes[0]) & 0x7f) + rawBytes[1:]  # Ensure slug starts with [A-Za-f]
+    rawBytes = bytearray(uuid.uuid4().bytes)
+    rawBytes[0] = rawBytes[0] & 0x7f;
```
Comment (petemoore) [inline slugid/slugid.py]: - I think we can lose the trailing `;`
- let's keep the comment to explain what this line does (`# Ensure slug starts with [A-Za-f]`)

File: test.py [hunk 3]
```diff
@@ -160,8 +160,16 @@ def spreadTest(generator, expected):
...
+    arraysEqual(expected, actual), "In a large sample of generated slugids, the range of characters found per character position in the sample did not match expected results.\n\nExpected: " + str(expected) + "\n\nActual: " + str(actual)
 
 def arraysEqual(a, b):
     """ returns True if arrays a and b are equal"""
-    return cmp(a, b) == 0
+    assert len(a) == len(b)
+    fails = False
+    msg = []
+    for x in range(0, len(a)):
+      if 0!= cmp(a[x], b[x]):
```
Comment (petemoore) [inline test.py]: - nit: can you add a space (` `) after the`0`?
- nit: I'd prefer `if cmp(a[x], b[x]) != 0` to `if 0 != cmp(a[x], b[x]):`

File: test.py [hunk 4]
```diff
@@ -160,8 +160,16 @@ def spreadTest(generator, expected):
         # sort for easy comparison
         actual[j] = ''.join(sorted(actual[j]))
 
-    assert arraysEqual(expected, actual), "In a large sample of generated slugids, the range of characters found per character position in the sample did not match expected results.\n\nExpected: " + str(expected) + "\n\nActual: " + str(actual)
+    arraysEqual(expected, actual), "In a large sample of generated slugids, the range of characters found per character position in the sample did not match expected results.\n\nExpected: " + str(expected) + "\n\nActual: " + str(actual)
 
 def arraysEqual(a, b):
     """ returns True if arrays a and b are equal"""
```
Comment (petemoore) [inline test.py:166]: Previously this was returning a boolean, but now the assert is inside - I prefer the method checking for equality rather than performing an assert inside.

File: test.py [hunk 5]
```diff
@@ -160,8 +160,16 @@ def spreadTest(generator, expected):
         # sort for easy comparison
         actual[j] = ''.join(sorted(actual[j]))
 
-    assert arraysEqual(expected, actual), "In a large sample of generated slugids, the range of characters found per character position in the sample did not match expected results.\n\nExpected: " + str(expected) + "\n\nActual: " + str(actual)
+    arraysEqual(expected, actual), "In a large sample of generated slugids, the range of characters found per character position in the sample did not match expected results.\n\nExpected: " + str(expected) + "\n\nActual: " + str(actual)
```
Comment (petemoore) [inline test.py]: arraysEqual no longer returns a boolean, so I'm not sure how this text gets used.
- What is the motivation of the change here? Is it to get more granular information about what the difference is between the two arrays?
- The previous implementation `return cmp(a, b) == 0` was pretty lightweight - did you find this inadequate, or is it related to python 3 compatibility?

**Comment by petemoore** [comment]
> I couldn't get the spread tests to work.  I also do not know how to make 'tox' run both the py35 and py25 environment tests...

I'd say let's merge when it is all working in python 3, rather than having broken tests.

**Comment by petemoore** [comment]
Note, when this is working in python 3, we should also update `classifiers` in setup.py to reflect this.

**Comment by petemoore** [comment]
P.S. And thanks for working on the python 3 support. :+1:

**Comment by jhford** [comment]
Thanks for the review.  I agree that we could wait until all the tests are passing to land it, but I'd like to get it landed sooner rather than later so that we don't block people from using Python 3 for downstream projects.  I think the two things we need fixed before we can merge are the tox.ini changes to test both py35 and py27 to run by default as well as the spread tests to be fixed.

I can take a stab at the tox changes shortly, but I found the spread tests to be a little confusing to read through.  Is there any chance that you might be able to tackle them?

File: test.py [hunk 6]
```diff
@@ -164,4 +164,10 @@ def spreadTest(generator, expected):
...
 def arraysEqual(a, b):
     """ returns True if arrays a and b are equal"""
-    return cmp(a, b) == 0
+    if len(a) != len(b):
+      return False
+    for x in range(0, len(a)):
+      if cmp(a[x], b[x]) != 0:
+        return False
+    return True
+
```
Comment (petemoore) [inline test.py:174]: Does `cmp(a, b)` not already do this, or does that break in python 3?

File: slugid/slugid.py [hunk 7]
```diff
@@ -38,6 +43,6 @@ def nice():
     Potentially other "nice" properties may be added in future to further
     restrict the range of potential uuids that may be generated.
     """
-    rawBytes = uuid.uuid4().bytes
-    rawBytes = chr(ord(rawBytes[0]) & 0x7f) + rawBytes[1:]  # Ensure slug starts with [A-Za-f]
+    rawBytes = bytearray(uuid.uuid4().bytes)
+    rawBytes[0] = rawBytes[0] & 0x7f
```
Comment (petemoore) [inline slugid/slugid.py:47]: That looks a lot, cleaner, thanks for that!

Comment (petemoore) [inline slugid/slugid.py:47]: Let's keep the comment `# Ensure slug starts with [A-Za-f]` though, I think it has value.

File: slugid/slugid.py [hunk 8]
```diff
@@ -1,8 +1,10 @@
 # Licensed under the Mozilla Public Licence 2.0.
 # https://www.mozilla.org/en-US/MPL/2.0
 
+import sys
 import uuid
 import base64
+two = True if sys.version_info.major == 2 else False
```
Comment (petemoore) [inline slugid/slugid.py:7]: Can we rename `two` to e.g. `isPython2`?

Alternatively as `sys.version_info.major` already holds this information, we could directly evaluate on line 21, e.g.

```
if sys.version_info.major == 2 and isinstance(slug, bytes):
```

Performance-wise I'm guessing the two options above are similar. I find the second option more readable than introducing an extra variable, (e.g. if python 4 came out, we'd potentially need `isPython3` and `isPython4` variables, whereas directly evaluating a condition against `sys.version_info.major` means we can already test if it is less/equal/bigger than a given value or satisfies a certain range etc. on a case-per-case basis).

**Comment by petemoore** [comment]
>  I agree that we could wait until all the tests are passing to land it, but I'd like to get it landed sooner rather than later so that we don't block people from using Python 3 for downstream projects.

That's a good point, thanks. Agreed.

**Comment by petemoore** [comment]
> I can take a stab at the tox changes shortly, but I found the spread tests to be a little confusing to read through. Is there any chance that you might be able to tackle them?

Sure, I'll have a stab.

**Comment by petemoore** [comment]
Closing #3, as this has been superseded by #4 (and the commits from this PR have been brought forward into #4).

File: slugid/slugid.py (hunk 8 above)
Comment (jhford) [inline slugid/slugid.py:7]: fwiw, performance is the same in python for this, but in JS, looking up members of an object is a performance drain because it has to go through the prototype chain each time.  I must have been thinking in JS when I wrote this :)
